// myshop/myshop.js
export var MyShopClient = new function(){
	this.about = { Version: 0.1, Author: "ML", Created: "", Updated: "-" };

	this.breakPoints = {
		xs: 768,
		sm: 992,
		md: 1200,
		lg: 1400,
		xlg: 1600
	};

	this.searchParams = new URLSearchParams(window.location.search);

	this.getShopId = function(){
		var shopId = (document.forms["data"]) ? ((document.forms["data"]["vid"]) ? document.forms["data"]["vid"].value : null) : null;
		shopId = (document.forms["data"]) ? ((document.forms["data"]["shopId"]) ? document.forms["data"]["shopId"].value : shopId) : shopId;
		shopId = (shopId == null) ? (this.searchParams.has('vid') ? (this.searchParams.get('vid')).toString() : null) : shopId;
		shopId = (shopId == null) ? (($("[name='vid']").val()) ? $("[name='vid']").val() : '5742100') : shopId;
		return shopId;
	};

	this.getCatalogId = function(){
		var cid = (document.forms["data"]) ? ((document.forms["data"]["cid"]) ? document.forms["data"]["cid"].value : null) : null;
		cid = (cid == null) ? (this.searchParams.has('id') ? (this.searchParams.get('id')).toString() : '0') : cid;
		return cid;
	};

	this.getProductId = function(){
		var pid = (document.forms["data"]) ? ((document.forms["data"]["pid"]) ? document.forms["data"]["pid"].value : null) : null;
		pid = (document.forms["data"]) ? ((document.forms["data"]["productid"]) ? document.forms["data"]["productid"].value : pid) : pid;
		pid = (pid == null) ? (this.searchParams.has('pid') ? (this.searchParams.get('pid')).toString() : '0') : pid;
		return pid;
	};

	this.getDataAttribute = function(vueRef, parameterName){
		var componentElement = vueRef;

		if(typeof vueRef == "object")
		{
			componentElement = vueRef.$root.$options.el;
			componentElement = componentElement.replace("#", "") + "-parent";
		}

		var el = document.getElementById(componentElement);
		return el.getAttribute('data-' + parameterName);
	};

	this.getLanguage = function(){
		var _lang = (document.forms["data"]) ? ((document.forms["data"]["_lang"]) ? document.forms["data"]["_lang"].value : null) : null;
		_lang = (_lang == null) ? ((this.searchParams.has('_lang') ? (this.searchParams.get('_lang')).toString() : 'nl')) : _lang;
		return _lang;
	};

	this.resourcesMap = null;
	this.loadResource = function(resourcesMapDefault, callbackFunction){
		var _self = this;
		if(this.resourcesMap == null)
		{
			this.resourcesMap = resourcesMapDefault;
			if(self.myshop){
				myshop("no").resources(function(me){
					if('getResource' in me)
					{
						_self.resourcesMap.forEach(function(currentValue, key, arr){
							var val1 = "" + me.getResource("", key);
							if('undefined' != val1 && "" != val1)
							{ 
								_self.resourcesMap.set(key, val1); 
							}
						});
					}
					callbackFunction();
				})
			}
			else{
				callbackFunction();
			}
		}
		else{
			callbackFunction();
		}
		return true;
	};

	this.getResource = function(key){
		if(this.resourcesMap !== null)
		{
			return this.resourcesMap.get(key);
		}
		return "";
	};

	/* Begin - OLD code for text Resources */
	// this.resource = null;
	// this.loadResource = function(callbackFunction){
	// 	if(this.resource == null)
	// 	{
	// 		var _self = this;
	// 		var xhr = new XMLHttpRequest();
	// 		xhr.open('GET', 'https://edit.myshop.com/resourceloader.json?lang=' + this.getLanguage());
	// 		xhr.onload = function() {
	// 		    if (xhr.status === 200) {
	// 		        _self.resource = JSON.parse(xhr.responseText);
	// 		        callbackFunction();
	// 		    }
	// 		    else {
	// 		        console.log('Request failed.  Returned status of ' + xhr.status);
	// 		    }
	// 		};
	// 		xhr.send();
	// 	}
	// 	else{
	// 		callbackFunction();
	// 	}
	// };
	// this.getResource = function(item, forceDefault){
	// 	var value = "";

	// 	if(this.resource != null)
	// 	{
	// 		var label = item.resourceDefault;

	// 		if(item.resourceId)
	// 			value = this.resource[item.resourceId];
	// 		if(forceDefault && "" != label)
	// 			value = label;
	// 		else if(!value && value != "" && label)
	// 			value = label;
	// 	}
	// 	return value;
	// };
	/* End - OLD code for text Resources */

	this.getRandomString = function (length) {
		length = (length) ? length : 5;
      	var text = "";
	  	var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

	  	for (var i = 0; i < length; i++)
	    	text += possible.charAt(Math.floor(Math.random() * possible.length));

	  	return text;
    };

    this.pad2 = function(n){ // always returns a string
    	return (n < 10 ? '0' : '') + n;
    };
    this.YYYYMMDDHHMMSS = function(date){
    	date = (date) ? date : new Date();
        return date.getFullYear() +
               this.pad2(date.getMonth() + 1) + 
               this.pad2(date.getDate()) +
               this.pad2(date.getHours()) +
               this.pad2(date.getMinutes()) +
               this.pad2(date.getSeconds());
    };

    this.getBreakPoints = function(){
    	return this.breakPoints;
    }

    this.flattenMe = function(a){
		var temp = [];
		if(a)
		for(var i = 0; i < a.length; i++){       
	        var children = [];
	        if(a[i].hasOwnProperty('children')){
	        	children = JSON.parse(JSON.stringify(a[i].children));
	            delete a[i].children;
	        }
	        temp.push(a[i]);
	    	if(children){
	        	if(children.length){
	            	var t = this.flattenMe(children);
	                temp = temp.concat(t);
	            }
	        }
	    }
	    
	    return temp;
	}
}
